pkg/server/plugin/upstreamauthority/spire/spire.go

```diff
@@ -70,59 +88,141 @@ func (m *Plugin) GetPluginInfo(context.Context, *plugin.GetPluginInfoRequest) (*
 	return &plugin.GetPluginInfoResponse{}, nil
 }
 
+func (m *Plugin) SetLogger(log hclog.Logger) {
+	m.pollMtx.Lock()
+	defer m.pollMtx.Unlock()
+	m.log = log.Named("upstream-authority-spire")
+}
+
 func (m *Plugin) MintX509CA(request *upstreamauthority.MintX509CARequest, stream upstreamauthority.UpstreamAuthority_MintX509CAServer) error {
-	m.mtx.RLock()
-	defer m.mtx.RUnlock()
-	ctx := stream.Context()
+	m.subscribeToPolling()
+	defer m.unsubscribeToPolling()
 
-	wCert, wKey, wBundle, err := m.getWorkloadSVID(ctx, m.config)
+	m.nodeMtx.RLock()
+	certChain, _, err := submitCSRUpstreamCA(stream.Context(), m.nodeClient, request.Csr)
+	m.nodeMtx.RUnlock()
 	if err != nil {
 		return err
 	}
 
-	conn, err := m.newNodeClientConn(ctx, wCert, wKey, wBundle)
-	if err != nil {
-		return err
+	ticker := time.NewTicker(internalPollFreq)
+	var rootCAs []*common.Certificate
+	for {
+		select {
+		case <-ticker.C:
+			newRootCAs := m.getBundle().RootCas
+			if !areRootsEqual(rootCAs, newRootCAs) {
```

Comment:
We have to be careful here. The first time through the roots could compare equal since the bundle may not have been fetched yet (both will be empty). If we're relying on the roots not being equal to send the first response, this may not do that.

More critically, we have to make sure the roots returned here is as fresh or fresher than the roots returned by the `submitCSRUpstreamCA` call.

Reply:
Good point! I'll update `m.currentBundle` with the rootsCA contained in the `submitCSRUpstreamCA` response.